airflow-core/src/airflow/utils/log/file_task_handler.py

```diff
@@ -180,15 +181,18 @@ def _interleave_logs(*logs: str | LogMessages) -> Iterable[StructuredLogMessage]
         last = msg
 
 
-def _ensure_ti(ti: TaskInstanceKey | TaskInstance, session) -> TaskInstance:
+def _ensure_ti(
+    ti: TaskInstanceKey | TaskInstance | TaskInstanceHistory, session
+) -> TaskInstance | TaskInstanceHistory:
     """
-    Given TI | TIKey, return a TI object.
+    Return a TaskInstance or TaskInstanceHistory for the given TI, TIKey, or TaskInstanceHistory.
 
-    Will raise exception if no TI is found in the database.
+    Raises AirflowException if no TI is found in the database.
     """
     from airflow.models.taskinstance import TaskInstance
+    from airflow.models.taskinstancehistory import TaskInstanceHistory
 
-    if isinstance(ti, TaskInstance):
+    if isinstance(ti, (TaskInstance, TaskInstanceHistory)):
         return ti
     val = (
```

Comment:
Is only querying TI the right behavior? Should this also fall back to history to get the right try? Just setting the try on the TI feels wrong.

Reply:
but yeah why is it setting try number 🤔

Reply:
This code path ensures we only hit the TaskInstanceHistory table when there’s no latest TaskInstance for the requested try.

Reply:
I have removed `_ensure_ti` helper and accepted `TaskInstanceHistory` in log handlers and reader.